Add Continuous Integration using Github Actions

cuminandpaprika opened the pull request:
Addresses https://github.com/anz-bank/sysl-go/issues/6

This PR implements some basic CI on the sysl-go repo
- Linting
   - It uses golangci-lint 1.23.8 for linting via the [Reviewdog Github Action](https://github.com/reviewdog/action-golangci-lint) as golangci.com is shutting down on the 15th April [blog post](https://medium.com/golangci/golangci-com-is-closing-d1fc1bd30e0e)
- Test Coverage
    - Runs `make check-coverage` to ensure test coverage is sufficient

Other steps explored in https://github.com/anz-bank/sysl-go/issues/6 are not addressed by this PR and will be added in subsequent PRs

juliaogris commented on .github/workflows/go.yml:
```diff
@@ -0,0 +1,49 @@
+        uses: actions/checkout@v2
+
+      - name: Get project dependencies
+        run: go get -v -t -d ./...
```
Same: `go get` is a pre go modules directive, you shouldn't have to run this.

juliaogris commented on .github/workflows/go.yml:
```diff
@@ -0,0 +1,49 @@
+      - name: Run tests
+        run: go test -cover -tags codeanalysis -coverprofile=coverage.out ./...
+
+      - name: Publish test results to Codecov
```
I think there is no need to publish this to Codecov. Do it with the makefile.

cuminandpaprika replied:
Yup, agreed, I've removed the dependency on Codecov (I think it just adds noise to the PR)

juliaogris commented on codecov.yml:
```diff
@@ -0,0 +1,6 @@
+  round: down
+  range: "50...100"
+ignore:
+  - codegen/tests/*
```
EOL on EOF.

juliaogris replied:
Here is why: https://gist.github.com/camh-/1bebfcff1b0f814e9b191edc60d5206b

cuminandpaprika replied:
Thanks, removed the codecov file

cuminandpaprika replied:
Fixed, thanks!

juliaogris reviewed (commented):
Ok, I've looked at this.
I'd say, avoid unnecessary external systems, especially codecov/coverall and linting (not sure about sonarcube).
I'd also say use the Makefile in the github actions as well so you can simulate a CI run locally with make.

juliaogris commented on .github/workflows/go.yml:
```diff
@@ -0,0 +1,49 @@
+        id: go
+
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@v2
```
Is it really necessary to checkout the code on every action "job".

cuminandpaprika replied:
I believe so, jobs run in parallel in separate environments so each one will need to checkout to code independently

Edit: Clearer explanation here https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#using-the-checkout-action

Docs here
https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobs

juliaogris replied:
Thanks.

cuminandpaprika commented:
> Ok, I've looked at this.
> I'd say, avoid unnecessary external systems, especially codecov/coverall and linting (not sure about sonarcube).
That's a good point, I've already run into some issues with some of the external services. I'll look into printing the codecoverage locally, thanks!

> I'd also say use the Makefile in the github actions as well so you can simulate a CI run locally with make.
Sounds good, definitely agree on the use of the Makefile to enable you to run as much as possible locally.

juliaogris commented on .golangci.yml:
```diff
@@ -0,0 +1,69 @@
+service:
+  golangci-lint-version: 1.19.x # use the fixed version to not introduce new linters unexpectedly
+  prepare:
+    - echo "here I can run custom commands, but no preparation needed for this repo"
```
Remove the line above.

cuminandpaprika replied:
Removed the entire service section, thanks!

juliaogris commented on .golangci.yml:
```diff
@@ -0,0 +1,69 @@
+      text: "ST1003:"
+      linters:
+        - stylecheck
+    # We like underscores
```
Really?

cuminandpaprika replied:
haha I just copied this over from the existing .golangci.yml file in `server-lib`, so can't comment on the accuracy of liking underscores :)

cuminandpaprika replied:
Removed this in the latest commits

cuminandpaprika commented:
@juliaogris 

Here's what I've gone with instead:
- Update the makefile so that CI can run make commands
- Uses [this](https://github.com/actions-contrib/golangci-lint ) action to run golangci-lint
    - I didn't see a need to make our own, and this one allows the version of golangci-lint to be specified
    - Potentially, something like https://github.com/marketplace/actions/golangci-lint-action would be great, as it posts annotations back to github, which link to the line of code that needs to be fixed. But it doesn't support specifying a version of golangci-lint
- Remove codecov and run `make test-coverage` as suggested

- [x] Still need to fix the linting issues

juliaogris commented on .golangci.yml:
```diff
@@ -0,0 +1,67 @@
+    - gocyclo
+    - goconst
+    - lll
+    - misspell
```
that's a good one to have imo.

In fact I'd keep:
lll, misspell, gochecknoglobals, and gocyclo.
You can update gocyclo to higher values in gocyclo specific settings;
you can also "exclude" individual lines / files with //nolint:lll for instance

cuminandpaprika replied:
That's a great point, 
I've updated to turn on `misspell`, `gocyclo` and `goconst`

It looks like it'd be quite a bit more work to enable `lll` and `gochecknoglobals` so I'll raise those as issues 
https://github.com/anz-bank/sysl-go/issues/33
https://github.com/anz-bank/sysl-go/issues/32

juliaogris commented on Makefile:
```diff
@@ -1,5 +1,52 @@
+tidy:
+	go mod tidy
+
+.PHONY: lint tidy clean
```
no clean target here?

cuminandpaprika replied:
Thanks, fixed in 728d9a010242a1d6784cb6665c4f590129a3b23a

juliaogris commented on validator/validator_test.go:
```diff
@@ -135,7 +135,7 @@ func testDummyValidator(reflect.Value) interface{} {
 
-//nolint:gochecknoinits
+//nolint:gochecknoinits // We ignore these for tests
 func init() {
```
Maybe this should / could be done in [TestMain](https://golang.org/pkg/testing/#hdr-Main)?

cuminandpaprika replied:
Good point, fixed in e97548cb87552a5b17071bd5097d429d82f62fe5

juliaogris commented on README.md:
```diff
@@ -14,19 +14,30 @@ Go get the repository
-- some working method of obtaining dependencies listed in `go.mod` (working internet access, `GOPROXY`)
-- env var `GOFLAGS="-mod=vendor"`
-- correctly configured cntlm or alpaca proxy running on localhost at port 3128 (only for updating vendor dependencies)
+- [go 1.13](https://golang.org/doc/install)
```
maybe time to upgrade to 1.14.1 ?

cuminandpaprika replied:
I think I'd like to keep the version of this at least consistent with Sysl, given that it's the same team of developers. Will create an issue for it though https://github.com/anz-bank/sysl-go/issues/34

juliaogris commented on Makefile:
```diff
@@ -1,5 +1,52 @@
+CHECK_COVERAGE = awk -F '[ \t%]+' '/^total:/ && $$3 < $(COVERAGE) {exit 1}'
+FAIL_COVERAGE = { echo '$(COLOUR_RED)FAIL - Coverage below $(COVERAGE)%$(COLOUR_NORMAL)'; exit 1; }
 
+.PHONY: check-coverage coverage test
```
`test` target missing.

cuminandpaprika replied:
Sorry, a little bit confused by this. The target for `test` is above I think
```
test: ## Run all tests
	go test -coverprofile=$(COVERFILE) -tags codeanalysis ./...
```

juliaogris replied:
yeah, so `.PHONY` needs to hold _all_ targets that don't match any actual file output generation. (this is an artefact from creating/ compiling c executables). Maybe we should talk about this at some point in more detail.

Anyway, long story short, you need to list `test` under `.PHONY` too.

cuminandpaprika replied:
Awesome, thanks for the explanation, `test` has been included undernder `.PHONY` in https://github.com/anz-bank/sysl-go/pull/23/commits/a459be5ee642cd9884c84430384c4e82f9e0abdd

juliaogris commented on Makefile:
```diff
@@ -1,5 +1,52 @@
+BOLD          = $(shell tput bold 2>/dev/null)
+
+help:
+	@awk -F ':.*## ' 'NF == 2 && $$1 ~ /^[A-Za-z0-9_-]+$$/ { printf "$(BOLD)$(COLOUR_WHITE)%-20s$(COLOUR_NORMAL)%s\n", $$1, $$2}' $(MAKEFILE_LIST) | sort
```
I think it's great that you are re-using this `help` target.

Maybe also add help messages for all targets with `##`.
You can look at the output with `make help`

cuminandpaprika replied:
Thanks for that, added help messages for all targets, it's a really nice feature to have in a makefile!

juliaogris reviewed (commented):
I think this is quite alright as is.
However, if you do agree with some comments and you don't want to update them as part of this PR, please write an issue for later.

It would also be great to have the git history cleaned up a bit.

juliaogris reviewed (commented):
This looks pretty good to me now, just needs that `.PHONY` update in the `Makefile`.
Also, would be great if you could squash your fixup commits and maybe just have three commits:
- linter updates
- makefile extension & docs update
- docs version updates
- CI
Or so....

cuminandpaprika committed:
Update Makefile

cuminandpaprika committed:
Fix linting issues

cuminandpaprika committed:
Add Github Action

cuminandpaprika committed:
Update docs for deps and development

cuminandpaprika committed:
Run go mod tidy

cuminandpaprika commented:
> This looks pretty good to me now, just needs that `.PHONY` update in the `Makefile`.
> Also, would be great if you could squash your fixup commits and maybe just have three commits:
> 
> * linter updates
> * makefile extension & docs update
> * docs version updates
> * CI
>   Or so....

Ah, thanks for the example, I wasn't sure what granularity is desirable for commits, that helps a lot to guide me!

Rebased and squashed, and addressed your comment on the makefile target 👍
